src/include/concurrency/decentralized_epoch_manager.h

```diff
@@ -0,0 +1,303 @@
+//===----------------------------------------------------------------------===//
+//
+//                         Peloton
+//
+// decentralized_epoch_manager.h
+//
+// Identification: src/include/concurrency/decentralized_epoch_manager.h
+//
+// Copyright (c) 2015-16, Carnegie Mellon University Database Group
+//
+//===----------------------------------------------------------------------===//
+
+
+#pragma once
+
+#include <thread>
+#include <vector>
+
+#include "common/macros.h"
+#include "type/types.h"
+#include "common/logger.h"
+#include "common/platform.h"
+#include "common/init.h"
+#include "common/thread_pool.h"
+#include "concurrency/epoch_manager.h"
+
+namespace peloton {
+namespace concurrency {
+
+
+class DecentralizedEpochManager : public EpochManager {
+  DecentralizedEpochManager(const DecentralizedEpochManager&) = delete;
+
+struct Epoch {
+  std::atomic<size_t> read_only_count_;
+  std::atomic<size_t> read_write_count_;
+
+  Epoch(): 
+    read_only_count_(0), 
+    read_write_count_(0) {}
+
+  Epoch(const Epoch &epoch): 
+    read_only_count_(epoch.read_only_count_.load()), 
+    read_write_count_(epoch.read_write_count_.load()) {}
+
+  void Init() {
+    read_only_count_ = 0;
+    read_write_count_ = 0;
+  }
+};
+
+struct LocalEpochContext {
+
+  LocalEpochContext(const size_t thread_id) : 
+    epoch_buffer_(epoch_buffer_size_),
+    head_epoch_id_(0),
+    tail_epoch_id_(UINT64_MAX),
+    thread_id_(thread_id) {}
+
+  bool EnterLocalEpoch(const uint64_t epoch_id) {
+    // if not initiated
+    if (tail_epoch_id_ == UINT64_MAX) {
+      tail_epoch_id_ = epoch_id - 1;
+    }
+
+    // ideally, epoch_id should never be smaller than head_epoch id.
+    // however, as we force updating head_epoch_id in GetTailEpochId() function,
+    // it is possible that we find that epoch_id is smaller than head_epoch_id.
+    // in this case, we should reject entering local epoch.
+    // this is essentially a validation scheme.
+    if (epoch_id < head_epoch_id_) {
+      return false;
+    }
+    
+    PL_ASSERT(epoch_id >= head_epoch_id_);
+
+    // update head epoch id.
+    head_epoch_id_ = epoch_id;
```

Comment:
Line 71 and line 78 should be done in an atomic way. Otherwise it will cause a race condition between GC thread and transaction worker.